### ISSUE TITLE
mdTooltip: $digest already in progress when there are a number of disabled elements

### ISSUE THREAD

**Issue by Frank3K** [angular/material#8011]
In Firefox 45.0.1 on Linux I encounter the following issue when loading a page with a number of disabled `md-button` elements, each with its own `md-tooltip`.

```
Error: [$rootScope:inprog] $digest already in progress
http://errors.angularjs.org/1.5.3/$rootScope/inprog?p0=%24digest
```

I have tracked down the issue to the following [line 157 in tooltip.js](https://github.com/angular/material/blob/e7f866d9e5f0239fe1b3b93f598bc1185966823d/src/components/tooltip/tooltip.js#L157):

``` js
      // add an mutationObserver when there is support for it
      // and the need for it in the form of viable host(parent[0])
      if (parent[0] && 'MutationObserver' in $window) {
        // use an mutationObserver to tackle #2602
        var attributeObserver = new MutationObserver(function(mutations) {
          mutations.forEach(function (mutation) {
            if (mutation.attributeName === 'disabled' && parent[0].disabled) {
              setVisible(false);
              scope.$digest(); // make sure the elements gets updated
            }
          });
        });

        attributeObserver.observe(parent[0], { attributes: true});
      }
```

When there are mutations for multiple elements (I enable most elements on a certain loaded event), a number of digest are executed if I read the code correctly. This is what (I think) triggers the error.

I'm no expert in `$digest()`, so I do not know what the best solution is. My first approach was to move the `scope.$digest()` outside the `forEach` but that did not fix the issue. Note that commenting the line altogether does remove the error, so I'm quite sure this is the offending call.

**Comment by clshortfuse** [comment]
I believe we can change scope.$digest() to angular.noop(). Do you have a codepen we can test with?

**Comment by Frank3K** [comment]
I don't have a codepen at the moment since the scenario seems quite complex (i.e. using a number of disabled elements that are enabled at some point in time). I will test your PR on my current project tonight.

**Comment by Frank3K** [comment]
Unfortunately the current fix from PR #8015 does not work. I'll try to construct a minimal example in a codepen.

**Comment by Frank3K** [comment]
I havent been able to reproduce the scenario in a codepen as of yet. Is it someway possible to get more information about the digest that is already in progress?

When looking at stack traces the conflicting digest is always the one I have pointed out in the first comment.

**Comment by Frank3K** [comment]
Theoretically, can one just trigger a digest inside a mutation observer (i.e. without a $timeout/$mdTick)? The mutation observer is outside the angular event loop.

Suppose the following scenario:

- A page contains multiple buttons with tooltips.
- For each tooltip a mutation observer is registered.
- The disabled state changes for multiple elements at once.
- I think the events are processed in sequence, say 1ms apart.
- If the digest phase takes more than one millisecond the 'digest already in progress' scenario is reached.

**Comment by topherfangio** [comment]
I don't have time to dig into this immediately, but we may be able to replace the `$scope.digest()` with the following:

    $mdUtil.nextTick(function() {
      setVisible(false);
    });

In theory, that should schedule the `setVisible()` call for the next digest loop instead of trying to fire it immediately.

@Frank3K Think you could give that a try?

**Comment by clshortfuse** [comment]
I think nextTick make more sense. I'm not sure about when digest gets processed, but it could be possible that the MutationObserver gets fired in the middle of a digest cycle.

Also, I just realized that even though changing the code would prevent multiple digests ever running in that loop because of the better use of Array.forEach vs Array.some, I think it's impossible that two attributes can have the name attributeName. All my code change does is short-circuit the evaluation and shave off a bit from the CPU

**Comment by Frank3K** [comment]
Thanks for your suggestions. I have combined both approaches and the following code eliminates the issue for my use case:

``` js
        var attributeObserver = new MutationObserver(function(mutations) {
          if (mutations.some(function (mutation) {
            return (mutation.attributeName === 'disabled' && parent[0].disabled);
          })) {
              $mdUtil.nextTick(function() {
                setVisible(false);
              });
            }
        });
```

@clshortfuse: can you update the existing PR?

**Comment by ThomasBurleson** [comment]
@Frank3K, @clshortfuse - nice work gentlemen! Thank you for your help with this.

**Comment by yoav-zibin** [comment]
Yes, I got it as well while using rc2 (I've now changed to rc4 --- hopefully it contains this fix :))

My stack trace:
Error: [$rootScope:inprog] http://errors.angularjs.org/1.5.3/$rootScope/inprog?p0=%24digest

http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:6:421
n@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:135:272
$apply@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:143:161
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:268:452
Pf@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:481
d@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:425
dispatchEvent@[native code]
o@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc2/angular-material.min.js:7:19474
http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc2/angular-material.min.js:7:15538
end@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc2/angular-material.min.js:7:19972
i@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc2/angular-material.min.js:7:20805
f@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc2/angular-material.min.js:7:21006
Pf@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:481
d@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:425
send@[native code]
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:101:265
n@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:97:165
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:94:310
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:128:306
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:17224
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:16024
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:17224
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:16024
$eval@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:142:467
$digest@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:140:52
$apply@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:143:256
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:3:1625




I'm currently only using md-fab and md-tooltip so it got to be the same issue:

<md-button id="test_open_new_match_modal" class="md-fab new-match-fab" aria-label="FAB"
      ng-click="main.showModal('newMatchModal')">
    <md-tooltip md-direction="left">
      {{ 'MAIN_NEW_MATCH_GROUP_TITLE' | translate }}
    </md-tooltip>
    <span style="font-size: 28px;">+</span>
  </md-button>

**Comment by yoav-zibin** [comment]
I still get it with rc4, is this fix in rc4 or not?

	Error: [$rootScope:inprog] http://errors.angularjs.org/1.5.3/$rootScope/inprog?p0=%24digest

http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:6:421
n@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:135:272
$apply@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:143:161
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:268:452
Pf@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:481
d@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:425
dispatchEvent@[native code]
o@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc4/angular-material.min.js:7:20522
http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc4/angular-material.min.js:7:16586
end@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc4/angular-material.min.js:7:21020
i@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc4/angular-material.min.js:7:21853
f@http://ajax.googleapis.com/ajax/libs/angular_material/1.1.0-rc4/angular-material.min.js:7:22054
Pf@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:481
d@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:36:425
send@[native code]
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:101:265
n@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:97:165
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:94:310
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:128:306
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:17224
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:16024
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:17224
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:1:16024
$eval@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:142:467
$digest@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:140:52
$apply@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:143:256
d@http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:3:7301
http://yoav-zibin.github.io/default-theme/app/index.html?phonegapPlugins=v100&onlyGameId=5125637257822208-1:4:13752
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:156:171
e@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:45:263
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.3/angular.min.js:48:177

**Comment by ThomasBurleson** [comment]
Can you reproduce the issue with Angular 1.5.5 or higher ?

**Comment by yoav-zibin** [comment]
I don't know how to reproduce it --- it happens [rarely] in my production code (and the error is reported back :))

I've upgraded to angular 1.5.5 and I'll post again if it happens.
Can you check if this fix is in your rc4 ?
